monai/transforms/spatial/functional.py

```diff
@@ -0,0 +1,131 @@
+    original_spatial_shape = img.peek_pending_shape() if isinstance(img, MetaTensor) else img.shape[1:]
+    src_affine: torch.Tensor = img.peek_pending_affine() if isinstance(img, MetaTensor) else torch.eye(4)
+    img = convert_to_tensor(data=img, track_meta=get_track_meta())
+    spatial_rank = min(len(img.shape) - 1, src_affine.shape[0] - 1, 3)
+    if (not isinstance(spatial_size, int) or spatial_size != -1) and spatial_size is not None:
+        spatial_rank = min(len(ensure_tuple(spatial_size)), 3)  # infer spatial rank based on spatial_size
+    src_affine = to_affine_nd(spatial_rank, src_affine).to(torch.float64)
+    dst_affine = to_affine_nd(spatial_rank, dst_affine) if dst_affine is not None else src_affine
+    dst_affine = convert_to_dst_type(dst_affine, src_affine)[0]
+    if not isinstance(dst_affine, torch.Tensor):
+        raise ValueError(f"dst_affine should be a torch.Tensor, got {type(dst_affine)}")
+
+    in_spatial_size = torch.tensor(original_spatial_shape[:spatial_rank])
+    if isinstance(spatial_size, int) and (spatial_size == -1):  # using the input spatial size
+        spatial_size = in_spatial_size
+    elif spatial_size is None and spatial_rank > 1:  # auto spatial size
+        spatial_size, _ = compute_shape_offset(in_spatial_size, src_affine, dst_affine)  # type: ignore
+    spatial_size = torch.tensor(fall_back_tuple(ensure_tuple(spatial_size)[:spatial_rank], in_spatial_size))
+    extra_info = {
+        "dtype": str(dtype_pt)[6:],  # remove "torch": torch.float32 -> float32
+        "mode": mode.value if isinstance(mode, Enum) else mode,
+        "padding_mode": padding_mode.value if isinstance(padding_mode, Enum) else padding_mode,
+        "align_corners": align_corners if align_corners is not None else TraceKeys.NONE,
+        "src_affine": src_affine,
+    }
+    try:
+        _s = convert_to_numpy(src_affine)
+        _d = convert_to_numpy(dst_affine)
+        xform = np.eye(spatial_rank + 1) if spatial_rank < 2 else np.linalg.solve(_s, _d)
+    except (np.linalg.LinAlgError, RuntimeError) as e:
+        raise ValueError(f"src affine is not invertible {_s}, {_d}.") from e
+    xform = convert_to_tensor(to_affine_nd(spatial_rank, xform)).to(device=img.device, dtype=torch.float64)
+    affine_unchanged = (
+        allclose(src_affine, dst_affine, atol=AFFINE_TOL) and allclose(spatial_size, in_spatial_size)
+    ) or (allclose(xform, np.eye(len(xform)), atol=AFFINE_TOL) and allclose(spatial_size, in_spatial_size))
+    lazy_evaluation = transform_info.get(TraceKeys.LAZY_EVALUATION, False)
+    meta_info = TraceableTransform.track_transform_meta(
+        img,
+        sp_size=spatial_size,
+        affine=None if affine_unchanged and not lazy_evaluation else xform,
+        extra_info=extra_info,
+        orig_size=original_spatial_shape,
+        transform_info=transform_info,
+        lazy_evaluation=lazy_evaluation,
+    )
+    img = img.as_tensor() if isinstance(img, MetaTensor) else img
```

Comment:
Could you please help me understand why we convert `MetaTensor` to `Tensor` here?
If must, can we execute it after line 102, then no need to convert it back to `MetaTensor`?

Thanks.

Reply:
I think line 102 is a shallow copy, so we should drop the current meta first. Skipping this will cause some bugs.

Reply:
Hi @KumoLiu ,

Could you please help add 1 line comment here to explain the code logic? Otherwise, I feel it's very confusing why we convert `MetaTensor` to `Tensor`.

Thanks in advance.

Reply:
Have added comments in this [commit](https://github.com/Project-MONAI/MONAI/pull/6031/commits/4281a597be03066743ac19577baabaf944f86d5c). Thanks!